File: packages/golang/options.go
package configdiscovery

// The functional-option surface. Every ambient input - the working directory, the home
// directory, the environment - has an option that replaces it, which is what makes the
// conformance probe and every test in this package hermetic.

type config struct {
	strategy   string
	arrayMerge string
	stopDir    string
	envPrefix  string
	profile    string
	strict     bool
	home       string
	cwd        string
	env        map[string]string
	envSet     bool
	defaults   map[string]any
	overrides  map[string]any
	relativeTo string
	warn       func(string)
}

// Option configures a Load call.
type Option func(*config)

// WithStrategy selects "layered" (the default) or "first-match" (SPEC section 3.2).
func WithStrategy(strategy string) Option { return func(c *config) { c.strategy = strategy } }

// WithArrayMerge selects "replace" (the default) or "concat" (SPEC section 4.3).
func WithArrayMerge(mode string) Option { return func(c *config) { c.arrayMerge = mode } }

// WithStopDir adds an inclusive stop condition to the upward walk (SPEC section 2.3).
func WithStopDir(dir string) Option { return func(c *config) { c.stopDir = dir } }

// WithEnvPrefix overrides the prefix for the environment layer (SPEC section 4.5).
func WithEnvPrefix(prefix string) Option { return func(c *config) { c.envPrefix = prefix } }

// WithProfile also loads config.<profile>.<ext> beside each base file (SPEC section 2.6).
func WithProfile(profile string) Option { return func(c *config) { c.profile = profile } }

// WithStrict promotes an unknown key from a warning to an error (SPEC section 5).
func WithStrict(strict bool) Option { return func(c *config) { c.strict = strict } }

// WithHome overrides the home directory the user-level root resolves under.
func WithHome(home string) Option { return func(c *config) { c.home = home } }

// WithCwd overrides the directory the upward walk starts from.
func WithCwd(cwd string) Option { return func(c *config) { c.cwd = cwd } }

// WithEnv replaces the environment the prefixed layer reads. An explicitly empty map means "no
// environment", which is different from not calling this option at all.
func WithEnv(env map[string]string) Option {
	return func(c *config) {
		c.env = env
		c.envSet = true
	}
}

// WithDefaults sets layer 0.
func WithDefaults(defaults map[string]any) Option {
	return func(c *config) { c.defaults = defaults }
}

// WithOverrides sets layer 5.
func WithOverrides(overrides map[string]any) Option {
	return func(c *config) { c.overrides = overrides }
}

// WithRelativeTo emits Source.Path relative to dir, forward-slashed. Used by the conformance
// probe; rarely useful otherwise.
func WithRelativeTo(dir string) Option { return func(c *config) { c.relativeTo = dir } }

// WithWarningHandler routes diagnostics. The default writes to stderr.
func WithWarningHandler(warn func(string)) Option { return func(c *config) { c.warn = warn } }

File: packages/rust/src/options.rs
//! The option surface - SPEC section 6.
//!
//! Every ambient input - the working directory, the home directory, the environment - has a
//! builder method that replaces it, which is what makes the conformance probe and every test in
//! this crate hermetic.

use std::collections::BTreeMap;
use std::path::PathBuf;

use serde_json::{Map, Value};

use crate::merge::{ArrayMerge, Strategy};

/// Everything `load` can be told.
pub struct Options {
    pub(crate) strategy: Strategy,
    pub(crate) array_merge: ArrayMerge,
    pub(crate) stop_dir: Option<PathBuf>,
    pub(crate) env_prefix: Option<String>,
    pub(crate) profile: Option<String>,
    pub(crate) strict: bool,
    pub(crate) home: Option<PathBuf>,
    pub(crate) cwd: Option<PathBuf>,
    pub(crate) env: Option<BTreeMap<String, String>>,
    pub(crate) defaults: Map<String, Value>,
    pub(crate) overrides: Map<String, Value>,
    pub(crate) relative_to: Option<PathBuf>,
    pub(crate) on_warning: Box<dyn Fn(&str)>,
}

impl std::fmt::Debug for Options {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Options")
            .field("strategy", &self.strategy)
            .field("array_merge", &self.array_merge)
            .field("stop_dir", &self.stop_dir)
            .field("env_prefix", &self.env_prefix)
            .field("profile", &self.profile)
            .field("strict", &self.strict)
            .field("home", &self.home)
            .field("cwd", &self.cwd)
            .field("relative_to", &self.relative_to)
            .finish_non_exhaustive()
    }
}

impl Default for Options {
    fn default() -> Self {
        Options {
            strategy: Strategy::default(),
            array_merge: ArrayMerge::default(),
            stop_dir: None,
            env_prefix: None,
            profile: None,
            strict: false,
            home: None,
            cwd: None,
            env: None,
            defaults: Map::new(),
            overrides: Map::new(),
            relative_to: None,
            // Diagnostics go to stderr, never stdout: PROBE.md section 4 depends on it, and so
            // does any consumer piping a program's output.
            on_warning: Box::new(|message| eprintln!("warning: {message}")),
        }
    }
}

impl Options {
    /// A fresh builder.
    pub fn builder() -> OptionsBuilder {
        OptionsBuilder {
            options: Options::default(),
        }
    }
}

/// Builds [`Options`].
#[derive(Debug, Default)]
pub struct OptionsBuilder {
    options: Options,
}

macro_rules! setter {
    ($(#[$meta:meta])* $name:ident, $field:ident, $type:ty) => {
        $(#[$meta])*
        #[must_use]
        pub fn $name(mut self, value: $type) -> Self {
            self.options.$field = value;
            self
        }
    };
}

impl OptionsBuilder {
    setter!(
        /// Whether every root contributes or only the nearest one with a file (SPEC section 3.2).
        strategy, strategy, Strategy
    );
    setter!(
        /// How arrays combine (SPEC section 4.3).
        array_merge, array_merge, ArrayMerge
    );
    setter!(
        /// Promote an unknown key from a warning to an error (SPEC section 5).
        strict, strict, bool
    );
    setter!(
        /// Layer 0.
        defaults, defaults, Map<String, Value>
    );
    setter!(
        /// Layer 5.
        overrides, overrides, Map<String, Value>
    );

    /// An extra, inclusive stop condition for the upward walk (SPEC section 2.3).
    #[must_use]
    pub fn stop_dir(mut self, dir: impl Into<PathBuf>) -> Self {
        self.options.stop_dir = Some(dir.into());
        self
    }

    /// Override the prefix for the environment layer (SPEC section 4.5).
    #[must_use]
    pub fn env_prefix(mut self, prefix: impl Into<String>) -> Self {
        self.options.env_prefix = Some(prefix.into());
        self
    }

    /// Also load `config.<profile>.<ext>` beside each base file (SPEC section 2.6).
    #[must_use]
    pub fn profile(mut self, profile: impl Into<String>) -> Self {
        self.options.profile = Some(profile.into());
        self
    }

    /// Override the home directory the user-level root resolves under.
    #[must_use]
    pub fn home(mut self, home: impl Into<PathBuf>) -> Self {
        self.options.home = Some(home.into());
        self
    }

    /// Override the directory the upward walk starts from.
    #[must_use]
    pub fn cwd(mut self, cwd: impl Into<PathBuf>) -> Self {
        self.options.cwd = Some(cwd.into());
        self
    }

    /// Replace the environment the prefixed layer reads. An explicitly empty map means "no
    /// environment", which is different from not calling this at all.
    #[must_use]
    pub fn env(mut self, env: BTreeMap<String, String>) -> Self {
        self.options.env = Some(env);
        self
    }

    /// Emit `Source::path` relative to `dir`, forward-slashed. Used by the conformance probe.
    #[must_use]
    pub fn relative_to(mut self, dir: impl Into<PathBuf>) -> Self {
        self.options.relative_to = Some(dir.into());
        self
    }

    /// Route diagnostics. The default writes to stderr.
    #[must_use]
    pub fn on_warning(mut self, warn: impl Fn(&str) + 'static) -> Self {
        self.options.on_warning = Box::new(warn);
        self
    }

    /// Finish.
    pub fn build(self) -> Options {
        self.options
    }
}
